### PR TITLE
ENH: `differentiate.jacobian`: support array `step_direction`

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-14946

#### What does this implement/fix?
https://github.com/scipy/scipy/issues/14946#issuecomment-2237508957 brought up that that it would be useful for the functions in `scipy.differentiate` to support bounds, that is, avoid evaluating the function outside a specified domain. 

`scipy.differentiate.derivative` (gh-21636), the function for elementwise scalar differentiation, already supported `step_direction` (i.e. one sided or central differences) and `initial_step` (the maximum step to take in the specified direction), which could be used to enforce bounds. 

`jacobian` already supports `initial_step`. This updates `jacobian` to accept `step_direction`. Together, they allow the user to avoid evaluating the function outside of specified bounds. 

#### Additional information
The interface could be wrapped to choose `initial_step` and `step_direction` based on `bounds` and the evaluation point `x`. However, step size and direction are fundamental concepts of finite different differentiation, so I think they should remain in the public interface.

### CODE REVIEW

**Review by j-bowhay — APPROVED** [review]
LGTM. Happy to merge once merge conflicts have been resolved

File: scipy/differentiate/_differentiate.py
Comment (mdhaber) [inline scipy/differentiate/_differentiate.py:908]:
```suggestion
    res = derivative(wrapped, x, tolerances=tolerances,
                     maxiter=maxiter, order=order, initial_step=initial_step,
                     step_factor=step_factor, preserve_shape=True,
                     step_direction=step_direction)
```

**Comment by mdhaber** [comment]
Thanks, Jake. I didn't mean to run the CI with `[docs only]`, but in retrospect, a full run probably isn't needed for the merge conflict fix since the only thing that changed is the function name. Tests pass locally. If you'd like me to re-run all tests, LMK, or if you're comfortable merging, sounds good to me. Thanks!

Next I think I would like to make this and `hessian` array API compatible. Should be easy because `derivative` is. One part of that would be to make `rosen` and its derivatives array-API compatible. Should that be a separate PR? Would you be interested in authoring that part (or reviewing)?

**Comment by j-bowhay** [comment]
Tests pass locally for me too so I think we are safe to merge given that CI was green before merging main

**Comment by j-bowhay** [comment]
> Next I think I would like to make this and `hessian` array API compatible. Should be easy because `derivative` is. One part of that would be to make `rosen` and its derivatives array-API compatible. Should that be a separate PR? Would you be interested in authoring that part? Reviewing?

Sounds like a good idea. I am away travelling from tomorrow for the next four months without a computer so I won't be able to author anything. I am happy to look at small and simple diffs on a mobile but anything longer or that might require playing with locally is going to be beyond reach for me.

**Pull request merged**
